templates/common/templatesource/upgrade_test.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2024 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package templatesource
+
+import (
+	"context"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+
+	abctestutil "github.com/abcxyz/abc/templates/testutil"
+	"github.com/abcxyz/pkg/testutil"
+)
+
+func TestForCanonical(t *testing.T) {
+	t.Parallel()
+
+	cases := []struct {
+		name                     string
+		location                 string
+		locType                  string
+		gitProtocol              string
+		destDir                  string
+		prependTempDirToLocation bool // set this for tests where location is a local path
+		dirContents              map[string]string
+		wantDownloader           Downloader
+		wantErr                  string
+	}{
+		{
+			name:        "remote_git_https_no_subdir",
+			location:    "github.com/abcxyz/abc",
+			locType:     "remote_git",
+			gitProtocol: "https",
+			wantDownloader: &remoteGitDownloader{
+				canonicalSource: "github.com/abcxyz/abc",
+				cloner:          &realCloner{},
+				remote:          "https://github.com/abcxyz/abc.git",
+				subdir:          "",
```

Comment:
Nits: Do we need to explicitly set subdir as "". Same below

Reply:
Fixed.